### PR TITLE
fix: issues from audit report

### PR DESCRIPTION
**Pull request by Pythonberg1997**
### Description

This pr is to fix some issues from audit report

### Changes

Notable changes:
* add more security check
* add `maxCallbackDataLength` to restrict callback data length
* remove `failReason` from `RetryPackage`
* add `_disableInitializers` to implementation contracts

### CODE REVIEW

File: contracts/middle-layer/resource-mirror/ObjectHub.sol [hunk 1]
```diff
@@ -12,13 +12,17 @@ import "../../interface/IObjectHub.sol";
...
+    /*----------------- initializer -----------------*/
     function initialize(address _ERC721_token, address _additional) public initializer {
-        ERC721Token = _ERC721_token;
-        additional = _additional;
+        __cmn_hub_init_unchained(_ERC721_token, _additional);
 
         channelId = OBJECT_CHANNEL_ID;
     }
 
+    function initializeV2() public reinitializer(2) {
```
Comment (guagualvcha) [inline contracts/middle-layer/resource-mirror/ObjectHub.sol:26]: The deployer should call `initializeV2`, dont they?

File: contracts/middle-layer/resource-mirror/CmnHub.sol [hunk 2]
```diff
@@ -169,6 +170,7 @@ abstract contract CmnHub is CmnStorage, Initializable, ICmnHub, IMiddleLayer {
                 bytes memory reason;
                 bool failed;
                 uint256 gasBefore = gasleft();
+                require(gasBefore > callbackGasLimit, "insufficient gas");
                 try
                     IApplication(extraData.appAddress).greenfieldCall{ gas: callbackGasLimit }(
```
Comment (guagualvcha) [inline contracts/middle-layer/resource-mirror/CmnHub.sol:194]: I think there is no need for this checker, if gas is not enough, the whole tx will be revert and cause no impact.

File: contracts/CrossChain.sol [hunk 3]
```diff
@@ -349,6 +349,14 @@ contract CrossChain is Config, Initializable, ICrossChain {
         emit Reopened(msg.sender);
     }
 
+    function emergencyChangeSequence(bool isSendSequence, uint8 channelId, bool isIncrease) external onlyEmergencyUpgradeOperator {
```
Comment (guagualvcha) [inline contracts/CrossChain.sol]: Please change to `emergencyChangeSequence(uint8 channelId, bool isSendSequence, bool isIncrease)`

Comment (cosinlink) [reply contracts/CrossChain.sol]: fixed

File: contracts/middle-layer/resource-mirror/ObjectHub.sol (hunk 1 above)
Comment (cosinlink) [reply contracts/middle-layer/resource-mirror/ObjectHub.sol:26]: the `Deployer` contract will call initializeV2 if it is deployed in the first time.

File: contracts/middle-layer/resource-mirror/CmnHub.sol (hunk 2 above)
Comment (cosinlink) [reply contracts/middle-layer/resource-mirror/CmnHub.sol:194]: It is under a whole try catch in the CrossChain contract, won't be reverted

**Review by guagualvcha — APPROVED** [review]

**Pull request merged**